sonic_data_client/events_client.go

```diff
@@ -0,0 +1,368 @@
+package client
+
+/*
+#cgo CFLAGS: -g -Wall -I../../sonic-swss-common/common -Wformat -Werror=format-security -fPIE 
+#cgo LDFLAGS: -L/usr/lib -lpthread -lboost_thread -lboost_system -lzmq -lboost_serialization -luuid -lswsscommon
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "events_wrap.h"
+*/
+import "C"
+
+import (
+    "strconv"
+    "fmt"
+    "reflect"
+    "sync"
+    "time"
+    "unsafe"
+
+    "github.com/go-redis/redis"
+
+    spb "github.com/sonic-net/sonic-gnmi/proto"
+    sdcfg "github.com/sonic-net/sonic-gnmi/sonic_db_config"
+    "github.com/Workiva/go-datastructures/queue"
+    log "github.com/golang/glog"
+    gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+const SUBSCRIBER_TIMEOUT = (2 * 1000)  // 2 seconds
+const EVENT_BUFFSZ = 4096
+
+const LATENCY_LIST_SIZE = 10  // Size of list of latencies.
+const PQ_MAX_SIZE = 10240     // Max cnt of pending events in PQ.
+
+// STATS counters
+const MISSED = "COUNTERS_EVENTS:missed_internal"
+const DROPPED = "COUNTERS_EVENTS:missed_by_slow_receiver"
+const LATENCY = "COUNTERS_EVENTS:latency_in_ms"
+
+var STATS_CUMULATIVE_KEYS = [...]string {MISSED, DROPPED}
+var STATS_ABSOLUTE_KEYS = [...]string {LATENCY}
+
+const STATS_FIELD_NAME = "value"
+
+const EVENTD_PUBLISHER_SOURCE = "{\"sonic-events-eventd"
+
+// Path parameter
+const PARAM_HEARTBEAT = "heartbeat"
+
+type EventClient struct {
+
+    prefix      *gnmipb.Path
+    path        *gnmipb.Path
+
+    q           *queue.PriorityQueue
+    channel     chan struct{}
+
+    wg          *sync.WaitGroup // wait for all sub go routines to finish
+
+    subs_handle unsafe.Pointer
+
+    stopped     int
+
+    // Stats counter
+    counters    map[string]uint64
+
+    last_latencies  [LATENCY_LIST_SIZE]uint64
+    last_latency_index  int
+    last_latency_full   bool
+
+    last_errors uint64
+}
+
+func set_heartbeat(val int) {
+    s := fmt.Sprintf("{\"HEARTBEAT_INTERVAL\":%d}", val)
+    rc := C.event_set_global_options(C.CString(s));
+    if rc != 0  {
+        log.V(4).Infof("Failed to set heartbeat val=%d rc=%d", val, rc)
+    }
+}
+
+func NewEventClient(paths []*gnmipb.Path, prefix *gnmipb.Path, logLevel int) (Client, error) {
+    var evtc EventClient
+    evtc.prefix = prefix
+    for _, path := range paths {
+        // Only one path is expected. Take the last if many
+        evtc.path = path
+    }
+
+    for _, e := range evtc.path.GetElem() {
+        keys := e.GetKey()
+        for k, v := range keys {
+            if (k == PARAM_HEARTBEAT) {
+                if val, err := strconv.Atoi(v); err == nil {
+                    log.V(7).Infof("evtc.heartbeat_interval is set to %d", val)
+                    set_heartbeat(val)
```

Comment:
break?

Reply:
This is parsing params passed in URL. As expected params is only one, it did not matter. Yet it is good to have a break.